### DISCUSSION TITLE
Advanced forms behaviours

### DISCUSSION THREAD

**Discussion by mjashanks** [Budibase/budibase#2033] [Ideas]
This is a collection of requested additions to forms. I may add to this.

### Select - load options from external data sources

This mentions loading with new SQL Connectors aka Datasource plus: https://github.com/Budibase/budibase/discussions/1877

I have also had the request to 
1. load options from a REST api
2. Apply a filter, based on other values on the form
3. Allow setting of a `Value` and a `Display` for the selected item

This also mentions filtering: https://github.com/Budibase/budibase/discussions/1617

We need to be able to trigger a filter, based on the value of a previous form field.

### Custom validation

Should be able to provide more options for validation. I suggest that this should be javascript based. We need to be able to define rules like:
`if 'Status' is 'Closed' then 'Closed Reason' is required`
`MinDate must be less than MaxDate`
`ClosedDate cannot be in the future`

https://github.com/Budibase/budibase/discussions/1679
https://github.com/Budibase/budibase/discussions/1795

This should also give the option to write custom error messages so that:
1. The user can write a better explanation e.g. "When you set the status to closed, you must provide a reason"
2. Translation to other languages (see below)

### Translation of validation messages

Currently, messages are always in english, users should be able to enter their own messages.

https://github.com/Budibase/budibase/discussions/1793
https://github.com/Budibase/budibase/discussions/2023
https://github.com/Budibase/budibase/discussions/1150

### Control width of Inputs

https://github.com/Budibase/budibase/discussions/1895

### Formatting of input values

e.g. allow currency symbols and thousands separators.

https://github.com/Budibase/budibase/discussions/1795

### Using datepickers with external datasources

https://github.com/Budibase/budibase/issues/1504

### Multipage forms

https://github.com/Budibase/budibase/discussions/1062

### Set default value on inputs

https://github.com/Budibase/budibase/discussions/1903

### Multi-select picker

https://github.com/Budibase/budibase/discussions/1863

### Clear form action

Ability to clear/reset a form on button click
https://github.com/Budibase/budibase/discussions/2043

### Edit Form in a modal

I realise that this is stretching the scope of this a bit, but thought I'd list this in here anyway as it is related: https://github.com/Budibase/budibase/discussions/1199

**Comment by KernOil** [comment]
I'm looking forward to these improvements this will really make my app work great!
